### PR TITLE
Move heavy coin selection out of the loop in SubmitDenominate

### PR DESCRIPTION
**Pull request by UdjinM6**
Have to refactor quite a bit of code but the functionality should stay the same (logic wise).

~Includes #2270 , will rebase after it's merged.~

### CODE REVIEW

File: src/privatesend-client.cpp [hunk 1]
```diff
@@ -1175,57 +1167,55 @@ void CPrivateSendClientManager::ProcessPendingDsaRequest(CConnman& connman)
...
+
     LOCK2(cs_main, pwalletMain->cs_wallet);
 
     std::string strError;
-    std::vector<CTxDSIn> vecTxDSInRet;
-    std::vector<CTxOut> vecTxOutRet;
+    std::vector< std::pair<CTxDSIn, CTxOut> > vecPSInOutPairs, vecPSInOutPairsTmp;
+
+    if (!SelectDenominate(strError, vecPSInOutPairs)) {
+        LogPrintf("CPrivateSendClientSession::SubmitDenominate -- SelectDenominate failed\n");
```
Comment (InhumanPerfection) [inline src/privatesend-client.cpp]: ... SelectDenominate failed, error: %s\n", strError);

File: src/privatesend-client.cpp [hunk 2]
```diff
@@ -1251,101 +1241,91 @@ bool CPrivateSendClientSession::PrepareDenominate(int nMinRounds, int nMaxRounds
...
+}
 
-    {
-        LOCK(pwalletMain->cs_wallet);
-        for (const auto& txin : vecTxDSIn) {
-            pwalletMain->LockCoin(txin.prevout);
-        }
+bool CPrivateSendClientSession::PrepareDenominate(int nMinRounds, int nMaxRounds, std::string& strErrorRet, const std::vector< std::pair<CTxDSIn, CTxOut> >& vecPSInOutPairsIn, std::vector< std::pair<CTxDSIn, CTxOut> >& vecPSInOutPairsRet)
+{
+    if (!pwalletMain) {
```
Comment (InhumanPerfection) [inline src/privatesend-client.cpp]: Same 3 checks at the beginning of SelectDenominate. Redundant here?

File: src/privatesend-client.cpp [hunk 3]
```diff
@@ -1251,101 +1241,91 @@ bool CPrivateSendClientSession::PrepareDenominate(int nMinRounds, int nMaxRounds
...
     int nStepsMax = 5 + GetRandInt(PRIVATESEND_ENTRY_MAX_SIZE - 5 + 1);
+    int nDenomResult{0};
+
+    std::vector<CAmount> vecStandardDenoms = CPrivateSend::GetStandardDenominations();
+    std::vector<int> vecSteps(vecStandardDenoms.size(), 0);
+    vecPSInOutPairsRet.clear();
 
-    while (nStep < nStepsMax) {
+    for (const auto& pair: vecPSInOutPairsIn) {
+        if (pair.second.nRounds < nMinRounds || pair.second.nRounds > nMaxRounds) {
```
[the remark this answers is not in the document]
Comment (UdjinM6) [reply src/privatesend-client.cpp]: Fixed by #2277

**Comment by UdjinM6** [comment]
rebased, pls review

**Pull request merged**